### PR TITLE
[FLINK-7479] [cep] Support to retrieve the past event by an offset

### PR DESCRIPTION
**Pull request by dianfu**
## What is the purpose of the change

*Currently, it's already able to retrieve events matched to the specifed pattern in IterativeCondition.Context. While there are also requirements to retrieve events by an physical offset. The retrieved events may not be matched to any pattern.*

## Brief change log

  - *Add API retain() in Pattern*
  - *Buffer the past events in NFA.process*
  - *Access the past events by the newly added API getEventByOffset in IterativeCondition.Context*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added test in IterativeConditionsITCase*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs / JavaDocs)

### CODE REVIEW

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 1]
```diff
@@ -269,9 +282,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * @param name Name of the new pattern
     * @return A new pattern which is appended to this one
     */
-  def next(name: String): Pattern[T, T] = {
-    Pattern[T, T](jPattern.next(name))
-  }
+  def next(name: String) = Pattern[T, T](jPattern.next(name))
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 2]
```diff
@@ -280,9 +291,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * @param name Name of the new pattern
     * @return A new pattern which is appended to this one
     */
-  def notNext(name: String): Pattern[T, T] = {
-    Pattern[T, T](jPattern.notNext(name))
-  }
+  def notNext(name: String) = Pattern[T, T](jPattern.notNext(name))
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 3]
```diff
@@ -292,9 +301,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * @param name Name of the new pattern
     * @return A new pattern which is appended to this one
     */
-  def followedBy(name: String): Pattern[T, T] = {
-    Pattern[T, T](jPattern.followedBy(name))
-  }
+  def followedBy(name: String) = Pattern[T, T](jPattern.followedBy(name))
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 4]
```diff
@@ -305,9 +312,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * @param name Name of the new pattern
     * @return A new pattern which is appended to this one
     */
-  def notFollowedBy(name : String): Pattern[T, T] = {
-    Pattern[T, T](jPattern.notFollowedBy(name))
-  }
+  def notFollowedBy(name : String) = Pattern[T, T](jPattern.notFollowedBy(name))
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 5]
```diff
@@ -317,9 +322,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * @param name Name of the new pattern
     * @return A new pattern which is appended to this one
     */
-  def followedByAny(name: String): Pattern[T, T] = {
-    Pattern[T, T](jPattern.followedByAny(name))
-  }
+  def followedByAny(name: String) = Pattern[T, T](jPattern.followedByAny(name))
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 6]
```diff
@@ -329,7 +332,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * @return The same pattern as optional.
     * @throws MalformedPatternException if the quantifier is not applicable to this pattern.
     */
-  def optional: Pattern[T, F] = {
+  def optional = {
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 7]
```diff
@@ -347,7 +350,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * @return The same pattern with a [[Quantifier.oneOrMore()]] quantifier applied.
     * @throws MalformedPatternException if the quantifier is not applicable to this pattern.
     */
-  def oneOrMore: Pattern[T, F] = {
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 8]
```diff
@@ -359,7 +362,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * @return The same pattern with number of times applied
     * @throws MalformedPatternException if the quantifier is not applicable to this pattern.
     */
-  def times(times: Int): Pattern[T, F] = {
+  def times(times: Int) = {
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 9]
```diff
@@ -372,7 +375,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * @return The same pattern with the number of times range applied
     * @throws MalformedPatternException if the quantifier is not applicable to this pattern.
     */
-  def times(from: Int, to: Int): Pattern[T, F] = {
+  def times(from: Int, to: Int) = {
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 10]
```diff
@@ -390,7 +393,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * @return The same pattern with the updated quantifier.
     * @throws MalformedPatternException if the quantifier is not applicable to this pattern.
     */
-  def allowCombinations(): Pattern[T, F] = {
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 11]
```diff
@@ -413,7 +416,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * By default a relaxed continuity is applied.
     * @return pattern with continuity changed to strict
     */
-  def consecutive(): Pattern[T, F] = {
+  def consecutive() = {
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 12]
```diff
@@ -426,7 +429,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * @param pattern the pattern to append
     * @return A new pattern which is appended to this one
     */
-  def followedBy(pattern: Pattern[T, F]): GroupPattern[T, F] =
+  def followedBy(pattern: Pattern[T, F]) =
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 13]
```diff
@@ -437,7 +440,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * @param pattern the pattern to append
     * @return A new pattern which is appended to this one
     */
-  def followedByAny(pattern: Pattern[T, F]): GroupPattern[T, F] =
+  def followedByAny(pattern: Pattern[T, F]) =
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

File: flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala [hunk 14]
```diff
@@ -449,7 +452,7 @@ class Pattern[T , F <: T](jPattern: JPattern[T, F]) {
     * @param pattern the pattern to append
     * @return A new pattern which is appended to this one
     */
-  def next(pattern: Pattern[T, F]): GroupPattern[T, F] =
+  def next(pattern: Pattern[T, F]) =
```
Comment (dawidwys) [inline flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala]: Unrelated change

**Comment by dianfu** [comment]
Thanks @dawidwys for the suggestion. I will update the PR accordingly.

**Comment by dianfu** [comment]
@dawidwys Have updated the PR according to your comments.
